[FIX] mail: show display_name for contact without name and email

kmdi-odoo opened the pull request:
Steps to reproduce
===============
1. Create a contact of type invoice address without name and email.
2. Go to any app with chatter.
3. Add this user to the recipient ----> Only the blue tick will be visible (recipient name will be empty)

After this commit, we will use the display_name as a fallback to show in the chatter.

seb-odoo commented:
@awa-odoo Ok so the problem seems bigger than anticipated.

So what these PR are telling is that we can have a partner with an empty name, but a display_name that has a value?

We previously had `nameOrDisplayName` on partner(/persona) model, at some point we removed it because we thought it was no longer necessary, but I guess we would need to re-introduce it everywhere we use `name` of partner in code then, which is a lot more than what is currently covered by these 2 PR.

awa-odoo commented:
Hello @seb-odoo 

I believe this is caused by the fact that we use res.partner records as "address" records, such as the described use case in the task  pad where you create an "invoice" type contact with no name. Those are not "real companies" nor "real persons".

I'm not sure every use case from mail/chatter makes a lot of sense in that context, for example you would most likely not have a livechat with such a contact, but Luc probably wanted to clean the few use cases where the interface shows "blank" or "false" (we have some extra cleaning outside of mail too).

If you'd prefer to handle that in your pipe, that's fine, just let me know (with a task link preferably) so I can do the proper followup 🙂 

Thanks for looking into it & cheers!

seb-odoo commented:
This is not a corner case, it can be broken in all flows, if the no-name partner is reachable in the UI it could be used everywhere (mention, follower, etc), and if it has an email, it could be author of messages, etc.

Just an example among many others
![image](https://github.com/user-attachments/assets/965c4141-a26c-4b17-baa2-e853373e99a3)

tde-banana-odoo commented:
Or we don't care for the 3 times this is going to appear. Having recipients without name and without email is a bad idea anyway.

awa-odoo commented on addons/mail/models/mail_thread.py:
```diff
@@ -1889,7 +1889,7 @@ def _message_add_suggested_recipient(self, result, partner=None, email=None, lan
             recipient_data.update({'partner_id': partner.id, 'name': partner.name or '', 'email': email_normalized})
         elif partner:  # incomplete profile: id, name
-            recipient_data.update({'partner_id': partner.id, 'name': partner.name})
+            recipient_data.update({'partner_id': partner.id, 'name': partner.name or partner.display_name})
```
@seb-odoo :wave: 

This one is a bit stalled, how can we make it move forward?
Just fix this part here and introduce another "display_name" key inside the dict then handle it on the JS side?

Is the rest ok?

cc @kmdi-odoo

kmdi-odoo commented:
Hello @awa-odoo :wave: 
cc @seb-odoo 

Updated the code and added tests.
Can you please have a look?

NB: There is a case where `display_name` itself will also be empty; when `type: invoice, parent: False, name: False, email: False`. But maybe that's too specific case to handle?

Thanks

kmdi-odoo commented on addons/mail/static/src/core/web/base_recipients_list.js:
```diff
@@ -21,8 +21,8 @@ export class BaseRecipientsList extends Component {
-            { partner }) => {
-                const text = partner.email ? partner.emailWithoutDomain : partner.name;
+            { partner, displayName }) => {
+                const text = partner.email ? partner.emailWithoutDomain : displayName;
```
Let's do `(partner.email && partner.emailWithoutDomain) || displayName` to be a little consistent with `SuggestedRecipientList`?
